### ISSUE TITLE
Request more training details

### ISSUE THREAD

**Issue by JamChaos** [lifuguan/IGGT_official#24]
Hi IGGT authors,
Thank you for your excellent work. I am currently trying to reproduce the results of the paper, but I have encountered a few implementation-related questions and would greatly appreciate your clarification.
**1.Pixel sampling strategy**
Could you please clarify how pixels are sampled during training?

Are they randomly sampled, or is there a specific sampling strategy (e.g., instance-balanced, hard mining, etc.)?

Approximately how many pixels are sampled per scene?

**2.Instance feature distance computation**
Regarding the instance contrastive loss:

Is the feature distance computed with or without the square root (i.e., L2 distance vs. squared L2 distance)?

Could you kindly provide the exact formula used for the pull and push terms?

Are the pull and push losses computed within each scene independently?

**3.Pull/Push loss ratio**
The paper appears to contain a minor typo regarding the pull/push weighting. “For hyper-parameter settings, we set λpull = 2.0, λpull = 1.0 and M = 1.0.” It seems that the second λpull should instead be λpush？From context, I inferred a 2:1 ratio, but training with this ratio does not reproduce the scale of the loss curves reported in the paper, and convergence behavior seems inconsistent.
Could you confirm the exact weighting between pull and push losses used in your experiments?

In my current reproduction, the model is able to roughly separate instances, but the results are not as clean or stable as those reported in the paper. Any additional training details would be extremely helpful.

Thank you very much for your time and assistance.

Best regards

**Comment by lifuguan** [comment]
Hi, 

Here are the loss code, hopes it can help:
```python

def multi_view_embedmask_contrastive_loss(
    features, instances, 
    delta_pull=0.25, delta_push=1.0, 
    delta_consistency=0.5, min_mask_pixels=50,
    weight_pull=2.0, weight_push=1.0, weight_consistency=2.0
):
    """
    Multi-view contrastive loss that ensures consistency across different views.

    Args:
        features: (B, N, C, H, W) tensor of pixel embeddings where N is number of views
        instances: list of length B, each element is (N, H, W, C_mask), C_mask is number of masks in that view
        delta_pull: margin for pull loss
        delta_push: margin for push loss  
        delta_consistency: margin for cross-view consistency loss
        min_mask_pixels: minimum number of pixels for a mask to be considered
        weight_pull: weight for pull loss
        weight_push: weight for push loss
        weight_consistency: weight for consistency loss

    Returns:
        tuple: (total_loss, pull_loss, push_loss, consistency_loss)
    """
    B, N, C, H, W = features.shape
    device = features.device

    total_pull_loss = 0.0
    total_push_loss = 0.0
    total_consistency_loss = 0.0

    # Store mask embeddings for each batch and mask index for consistency computation
    # Structure: {b: {mask_idx: [mean_feat_view0, mean_feat_view1, ...]}}
    batch_mask_embeddings = {}

    for b in range(B):
        batch_mask_embeddings[b] = {}
        # For each view
        for view_idx in range(N):
            feat = features[b, view_idx]  # (C, H, W)
            feat = feat.permute(1, 2, 0).reshape(-1, C)  # (H*W, C)

            # instances[b] shape: (N, H, W, C_mask)
            # For this view, get (H, W, C_mask)
            masks_view = instances[b][view_idx]  # (H, W, C_mask)
            if masks_view is None or masks_view.numel() == 0:
                continue
            if masks_view.dim() == 2:  # (H, W) single mask, unsqueeze to (H, W, 1)
                masks_view = masks_view.unsqueeze(-1)
            Hm, Wm, C_mask = masks_view.shape
            if Hm != H or Wm != W:
                # Resize masks to (H, W)
                masks_view = resize_masks(masks_view.permute(2, 0, 1), (H, W)).permute(1, 2, 0)  # (H, W, C_mask)
                Hm, Wm, C_mask = masks_view.shape

            instance_means = []
            pull_loss = 0.0
            valid_mask_count = 0

            for mask_idx in range(C_mask):
                mask = masks_view[..., mask_idx]  # (H, W), bool or 0/1
                mask = (mask > 0)  # Only keep valid region
                mask_flat = mask.view(-1)
                if mask_flat.sum() < min_mask_pixels:
                    continue  # Skip tiny instances

                pixel_feats = feat[mask_flat]  # (num_pixels, C)
                if pixel_feats.shape[0] == 0:
                    continue
                mean_feat = pixel_feats.mean(dim=0, keepdim=True)  # (1, C)

                # Pull loss: pixels within same mask should be close
                pull = torch.sqrt(((pixel_feats - mean_feat) ** 2).sum(dim=1))  # (num_pixels,)
                pull = F.relu(pull - delta_pull).mean()
                pull_loss += pull
                valid_mask_count += 1

                # Store mean embedding for consistency loss
                if mask_idx not in batch_mask_embeddings[b]:
                    batch_mask_embeddings[b][mask_idx] = []
                batch_mask_embeddings[b][mask_idx].append(mean_feat.squeeze(0))  # (C,)

                instance_means.append(mean_feat)

            if valid_mask_count > 0:
                total_pull_loss += pull_loss / valid_mask_count

            # Push loss: different masks should be far apart
            if len(instance_means) > 1:
                means = torch.cat(instance_means, dim=0)  # (num_valid_masks, C)
                dist_mat = torch.cdist(means, means, p=2)  # (num_valid_masks, num_valid_masks)
                eye = torch.eye(len(means), device=device).bool()
                push_loss = F.relu(delta_push - dist_mat)[~eye].mean()
                total_push_loss += push_loss

    # Second pass: compute consistency loss across views
    for b in range(B):
        for mask_idx, view_embeddings in batch_mask_embeddings[b].items():
            if len(view_embeddings) < 2:
                continue  # Need at least 2 views for consistency
            mask_embeddings = torch.stack(view_embeddings, dim=0)  # (num_views, C)
            dist_mat = torch.cdist(mask_embeddings, mask_embeddings, p=2)  # (num_views, num_views)
            eye = torch.eye(len(mask_embeddings), device=device).bool()
            consistency_loss = F.relu(dist_mat[~eye] - delta_consistency).mean()
            total_consistency_loss += consistency_loss

    # Average losses
    avg_pull = total_pull_loss / (B * N) if B * N > 0 else 0
    avg_push = total_push_loss / (B * N) if B * N > 0 else 0

    # Count total mask instances across all batches and views
    total_mask_instances = sum(
        1 for batch_embeddings in batch_mask_embeddings.values()
        for mask_embeddings in batch_embeddings.values() if len(mask_embeddings) >= 2
    )
    avg_consistency = total_consistency_loss / total_mask_instances if total_mask_instances > 0 else 0

    # Combine losses with weights
    total_loss = (
        weight_pull * avg_pull +
        weight_push * avg_push +
        weight_consistency * avg_consistency
    )

    return total_loss, avg_pull, avg_push, avg_consistency
```

**Comment by JamChaos** [comment]
Thank you very much for your detailed clarification. I really appreciate your time and helpful explanations. This is very helpful for my reproduction!

**Comment by JamChaos** [comment]
Hi,

I have one additional question regarding the contrastive loss implementation.

Before computing the contrastive loss, are the feature vectors L2-normalized?
In addition, when computing the prototype as:
mean_feat = pixel_feats.mean(dim=0, keepdim=True)  # (1, C)
Does the mean feature not need to be normalized after averaging?

**Comment by lifuguan** [comment]
Yes, the feature is normalized by L2. You can directly use this function below:
```python
            stu_predictions['part_feat'] = F.normalize(stu_predictions['part_feat'], p=2, dim=2)
            # Extract instances from batch & Compute contrastive loss using instances
            final_loss, avg_pull, avg_push, avg_consistency = multi_view_embedmask_contrastive_loss(stu_predictions['part_feat'], masks_gt)

```

**Comment by JamChaos** [comment]
Thanks again, I successfully reproduced the result!

**Comment by PhucNDA** [comment]
Hi @JamChaos,

Could you share more on the implementation details ? I also plan to reproduce the result from the paper

Best,
PhucNDA.
